File: basics/Sorting/PHP/MergeSort.php
<?php
/*
 * Basic example of Merge Sort
 * Divide-and-conquer alrorithm
 * Complexity: O (n log n)
 */

function mergeSort($dataset){

	if(count($dataset) == 1 ) {
		return $dataset;
	}

	$middle = count($dataset) / 2;

    $leftset  = array_slice($dataset, 0, $middle);
    $rightset = array_slice($dataset, $middle);

	$leftset  = mergeSort($leftset);
	$rightset = mergeSort($rightset);

	$result = array();

	while (count($leftset) > 0 && count($rightset) > 0){
		if($leftset[0] > $rightset[0]){
			$result[] = $rightset[0];
			$rightset = array_slice($rightset , 1);
		}else{
			$result[] = $leftset[0];
			$leftset = array_slice($leftset, 1);
		}
	}

	while (count($leftset) > 0){
		$result[] = $leftset[0];
		$leftset = array_slice($leftset, 1);
	}

	while (count($rightset) > 0){
		$result[] = $rightset[0];
		$rightset = array_slice($rightset, 1);
	}

	return $result;
}


$array = [23, 8, 15, 17, 4, 40, 11, 31, 55, 3, 1, 100,99];
$result = mergeSort($array);

echo "Original: " . json_encode($array) . "\n";
echo "Result:   " . json_encode($result) . "\n";

File: basics/Sorting/Python/MergeSort.py
"""
Basic example of Merge Sort
Divide-and-conquer alrorithm

Complexity: O (n log n)
"""

def mergeSort(dataset):
    if len(dataset) > 1:
        mid = len(dataset) // 2
        leftarr = dataset[:mid]
        rightarr = dataset[mid:]

        # recursively break down the arrays
        mergeSort(leftarr)
        mergeSort(rightarr)

        # now perform the merging
        i=0 # index into the left array
        j=0 # index into the right array
        k=0 # index into merged array

        # while both arrays have content
        while i < len(leftarr) and j < len(rightarr):
            if leftarr[i] < rightarr[j]:
                dataset[k] = leftarr[i]
                i += 1
            else:
                dataset[k] = rightarr[j]
                j += 1
            k += 1

        # if the left array still has values, add them
        while i < len(leftarr):
            dataset[k] = leftarr[i]
            i += 1
            k += 1

        # if the right array still has values, add them
        while j < len(rightarr):
            dataset[k] = rightarr[j]
            j += 1
            k += 1


def main():
	items = [6, 20, 8, 19, 56, 23, 87, 41, 49, 53]
	# test the merge sort with data
	print(f"Original list: {items}")
	mergeSort(items)
	print(f"Sorted list:   {items}")

if __name__ == "__main__":
	main()
